src/libraries/Native/Unix/System.Native/pal_runtimeinformation.c

```diff
@@ -58,32 +61,72 @@ int32_t SystemNative_GetUnixVersion(char* version, int* capacity)
  1 - x64
  2 - ARM
  3 - ARM64
- 4 - WASM */
+ 4 - WASM
+ 5 - S390X */
 int32_t SystemNative_GetOSArchitecture()
 {
-#if defined(TARGET_ARM)
-    return ARCH_ARM;
-#elif defined(TARGET_ARM64)
-    return ARCH_ARM64;
-#elif defined(TARGET_AMD64)
-    return ARCH_X64;
-#elif defined(TARGET_X86)
-    return ARCH_X86;
-#elif defined(TARGET_WASM)
+#ifdef TARGET_WASM
     return ARCH_WASM;
-#elif defined(TARGET_S390X)
-    return ARCH_S390X;
 #else
-#error Unidentified Architecture
+    int32_t result = -1;
+#ifdef TARGET_SUNOS
+    // On illumos/Solaris, the recommended way to obtain machine
+    // architecture is using `sysinfo` rather than `utsname.machine`.
+
+    char isa[32];
+    if (sysinfo(SI_ARCHITECTURE_K, isa, sizeof(isa)) > -1)
+    {
+#else
+    struct utsname _utsname;
+    if (uname(&_utsname) > -1)
+    {
+        char* isa = _utsname.machine;
+#endif
+        // aarch64 or arm64: arm64
+        if (strcmp("aarch64", isa) == 0 || strcmp("arm64", isa) == 0)
+        {
+            result = ARCH_ARM64;
+        }
+
+        // starts with "arm": arm
+        else if (strncmp("arm", isa, strlen("arm")) == 0)
+        {
+            result = ARCH_ARM;
+        }
+
+        // x86_64 or amd64: x64
+        else if (strcmp("x86_64", isa) == 0 ||
+             strcmp("amd64", isa) == 0)
+        {
+            result = ARCH_X64;
+        }
+
+        // ix86: x86
+        else if (strncmp("i", isa, strlen("i")) == 0 &&
+             strncmp("86", isa + strlen("ix"), strlen("86")) == 0)
+        {
+            result = ARCH_X86;
+        }
+
+        else if (strcmp("s390x", isa) == 0)
+        {
+             result = ARCH_S390X;
+        }
+    }
+
+    // catch if we have missed a pattern above.
+    assert(result != -1);
+    return result;
```

Comment:
If it was me, I would have it throw as fallback and force them to create issue to let someone know about the missing OS Arch.

Reply:
> If it was me, I would have it throw as fallback and force them to create issue to let someone know about the missing OS Arch.

I don't think people would expect that, I think they'd just rather it says x86 or the processor architecture - I know this is certainly what I'd expect (the issue would be in the future if you're running older dotnet apps on newer architectures that don't exist / aren't supported by this .NET version natively), but it could be good to have a way of indicating it is unknown (I think possibly the processor architecture may be a better fallback, but in my code I use x86 as fallback so...)